udp/client/clientconn.go

```diff
@@ -613,113 +673,31 @@ func (cc *ClientConn) Process(datagram []byte) error {
 	cc.CheckMyMessageID(req)
 	cc.inactivityMonitor.Notify()
 	err = cc.goPool(func() {
-		defer cc.inactivityMonitor.Notify()
-		reqMid := req.MessageID()
-
-		// The same message ID can not be handled concurrently
-		// for deduplication to work
-		l := cc.msgIdMutex.Lock(reqMid)
-		defer l.Unlock()
-
-		origResp := cc.AcquireMessage(cc.Context())
-		origResp.SetToken(req.Token())
-		// If a request is sent in a Non-confirmable message, then the response
-		// is sent using a new Non-confirmable message, although the server may
-		// instead send a Confirmable message.
-		origResp.SetType(req.Type())
-		w := NewResponseWriter(origResp, cc, req.Options())
-		if ok, err := cc.getResponseFromCache(req.MessageID(), w.response); ok {
-			defer cc.ReleaseMessage(w.response)
+		defer func() {
 			if !req.IsHijacked() {
-				defer cc.ReleaseMessage(req)
-			}
-			if req.Type() == udpMessage.Confirmable {
-				w.response.SetType(udpMessage.Acknowledgement)
-				w.response.SetMessageID(reqMid)
-			} else {
-				w.response.SetType(udpMessage.NonConfirmable)
-				w.response.SetMessageID(cc.getMID())
-			}
-			err = cc.session.WriteMessage(w.response)
-			if err != nil {
-				closeConnection()
-				cc.errors(fmt.Errorf("cannot write response: %w", err))
-				return
+				cc.ReleaseMessage(req)
 			}
-			return
-		} else if err != nil {
+		}()
+		resp := cc.AcquireMessage(cc.Context())
+		resp.SetToken(req.Token())
```

Comment:
Previously
```
origResp.SetType(req.Type())
```

was called after `SetToken`, but the call to `SetType` is no longer here. I see that `cc.processReq` calls `w.response.SetType` in almost all code branches except for inner `if` for the case `w.response.IsModified() && (w.response.Type() == udpMessage.Reset || w.response.Code() == codes.Empty)`. So for that code path the response type might be different, is that OK?

Reply:
Also originally for `cc.getResponseFromCache(req.MessageID(), w.response); ok` with `ok = true` the code was 

```go
if req.Type() == udpMessage.Confirmable {
  w.response.SetType(udpMessage.Acknowledgement)
  w.response.SetMessageID(reqMid)
} else {
  w.response.SetType(udpMessage.NonConfirmable)
  w.response.SetMessageID(cc.getMID())
}
```

The `SetType` call is now missing in the else branch, is that OK?

Reply:
Fixed